File: enode/src/main/java/org/enodeframework/eventing/EventAppendResult.java
package org.enodeframework.eventing;

import com.google.common.collect.Lists;
import com.google.common.collect.Maps;

import java.util.List;
import java.util.Map;

public class EventAppendResult {

    private final Object lockObj = new Object();

    private List<String> successAggregateRootIdList = Lists.newArrayList();

    private List<String> duplicateEventAggregateRootIdList = Lists.newArrayList();

    private Map<String, List<String>> duplicateCommandAggregateRootIdList = Maps.newHashMap();

    public List<String> getSuccessAggregateRootIdList() {
        return successAggregateRootIdList;
    }

    public void setSuccessAggregateRootIdList(List<String> successAggregateRootIdList) {
        this.successAggregateRootIdList = successAggregateRootIdList;
    }

    public List<String> getDuplicateEventAggregateRootIdList() {
        return duplicateEventAggregateRootIdList;
    }

    public void setDuplicateEventAggregateRootIdList(List<String> duplicateEventAggregateRootIdList) {
        this.duplicateEventAggregateRootIdList = duplicateEventAggregateRootIdList;
    }

    public Map<String, List<String>> getDuplicateCommandAggregateRootIdList() {
        return duplicateCommandAggregateRootIdList;
    }

    public void setDuplicateCommandAggregateRootIdList(Map<String, List<String>> duplicateCommandAggregateRootIdList) {
        this.duplicateCommandAggregateRootIdList = duplicateCommandAggregateRootIdList;
    }

    public void addSuccessAggregateRootId(String aggregateRootId) {
        synchronized (lockObj) {
            if (!successAggregateRootIdList.contains(aggregateRootId)) {
                successAggregateRootIdList.add(aggregateRootId);
            }
        }
    }

    public void addDuplicateEventAggregateRootId(String aggregateRootId) {
        synchronized (lockObj) {
            if (!duplicateEventAggregateRootIdList.contains(aggregateRootId)) {
                duplicateEventAggregateRootIdList.add(aggregateRootId);
            }
        }
    }

    public void addDuplicateCommandIds(String aggregateRootId, List<String> aggregateDuplicateCommandIdList) {
        synchronized (lockObj) {
            if (!duplicateCommandAggregateRootIdList.containsKey(aggregateRootId)) {
                duplicateCommandAggregateRootIdList.put(aggregateRootId, aggregateDuplicateCommandIdList);
            }
        }
    }
}

File: enode/src/main/java/org/enodeframework/eventing/EventAppendResult.kt
package org.enodeframework.eventing

import com.google.common.collect.Lists
import com.google.common.collect.Maps

class EventAppendResult {
    private val lockObj = Any()
    var successAggregateRootIdList: MutableList<String> = Lists.newArrayList()
    var duplicateEventAggregateRootIdList: MutableList<String> = Lists.newArrayList()
    var duplicateCommandAggregateRootIdList: MutableMap<String, List<String>> = Maps.newHashMap()

    fun addSuccessAggregateRootId(aggregateRootId: String) {
        synchronized(lockObj) {
            if (!successAggregateRootIdList.contains(aggregateRootId)) {
                successAggregateRootIdList.add(aggregateRootId)
            }
        }
    }

    fun addDuplicateEventAggregateRootId(aggregateRootId: String) {
        synchronized(lockObj) {
            if (!duplicateEventAggregateRootIdList.contains(aggregateRootId)) {
                duplicateEventAggregateRootIdList.add(aggregateRootId)
            }
        }
    }

    fun addDuplicateCommandIds(aggregateRootId: String, aggregateDuplicateCommandIdList: List<String>) {
        synchronized(lockObj) {
            if (!duplicateCommandAggregateRootIdList.containsKey(aggregateRootId)) {
                duplicateCommandAggregateRootIdList[aggregateRootId] = aggregateDuplicateCommandIdList
            }
        }
    }
}
